### PR TITLE
Add rp_pico plugin

### PR DESCRIPTION
**Pull request by zeha**
Mostly soliciting feedback for now. This branch includes hughsie/subsystem_devtype and zeha/uf2-reenum.

Type of pull request:

- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

plugin checklist:

- [x] Fill out `README.md` with update protocol
- [x] Fill out `README.md` with any custom quirks and flags
- [ ] Fill out `README.md` with the vendor ID security value
- [ ] ~~Implement `FuFirmware->write()` and include at least one fuzzer testcase in `src/fuzzing/firmware` for any custom `FuFirmware` subclass~~
- [ ] CI run of the plugin for at least one target
- [ ] Document targets CI isn't currently run and the reasons (i.e. minimum versions needed or distribution limitations etc).

### CODE REVIEW

File: plugins/mntre/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+## Update Behavior
+
+The device is updated by triggering a reset, and then entering the Raspberry Pi Pico bootrom.
+
+The bootrom speaks the UF2 protocol.
+
+## Version Considerations
+
+This plugin has been available since fwupd version `2.0.SET_VERSION_HERE`.
```
Comment (hughsie) [inline plugins/mntre/README.md:27]: use 2.0.2 for now, we can fixup if it misses the release.

File: plugins/mntre/fu-mntre-device.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+				     struct mntre_reset_firmware_version *firmware_version)
+{
+	g_autoptr(GError) error_local = NULL;
+	gsize actual_len = 0;
+
+	guint16 reset_interface = 2; // TODO
+
+	FuUsbDeviceClaimFlags flags = 0;
+
+	if (!fu_usb_device_claim_interface(FU_USB_DEVICE(self), reset_interface, flags, error)) {
```
Comment (hughsie) [inline plugins/mntre/fu-mntre-device.c:32]: this is normally done in _setup automatically -- can you try using `fu_usb_device_add_interface(FU_USB_DEVICE(self), 2)` in __init() and removing this?

File: plugins/mntre/fu-mntre-device.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	return TRUE;
+}
+
+static gboolean
+fu_mntre_device_ensure_version(FuMntreDevice *self, GError **error)
+{
+	g_autofree gchar *version = NULL;
+	struct mntre_reset_firmware_version firmware_version = {};
+
+	if (!fu_mntre_device_get_firmware_version(self, error, &firmware_version))
```
Comment (hughsie) [inline plugins/mntre/fu-mntre-device.c:117]: ahh, rename this fu_mntre_device_ensure_firmware_version() and then call fu_device_set_version_raw() in that with the firmware_version.version

File: plugins/mntre/fu-mntre-device.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+		return FALSE;
+
+	version = g_strdup_printf("%08u", firmware_version.version);
+
+	fu_device_set_version(FU_DEVICE(self), version);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_NUMBER);
```
Comment (hughsie) [inline plugins/mntre/fu-mntre-device.c:123]: this should be set in _init()

File: plugins/mntre/fu-mntre-device.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+{
+	g_autofree gchar *version = NULL;
+	struct mntre_reset_firmware_version firmware_version = {};
+
+	if (!fu_mntre_device_get_firmware_version(self, error, &firmware_version))
+		return FALSE;
+
+	version = g_strdup_printf("%08u", firmware_version.version);
+
+	fu_device_set_version(FU_DEVICE(self), version);
```
Comment (hughsie) [inline plugins/mntre/fu-mntre-device.c:122]: Can you use FuDeviceClass->convert_version() instead pls

File: plugins/mntre/fu-mntre-device.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	FuMntreDevice *self = FU_MNTRE_DEVICE(device);
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_mntre_device_parent_class)->setup(device, error)) {
+		return FALSE;
+	}
+
+	if (!fu_mntre_device_ensure_version(self, error)) {
+		return FALSE;
+	}
```
Comment (hughsie) [inline plugins/mntre/fu-mntre-device.c:140]: no {} here pls.

File: plugins/mntre/mntre_reset.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+#define MNTRE_RESET_REQUEST_RESET 0x02
+// read current firmware version. Returns mntre_reset_firmware_version.
+#define MNTRE_RESET_GET_FIRMWARE_VERSION 0x03
+
+// String name of the interface
+#define MNTRE_RESET_INTERFACE_NAME_STR "Reset"
+
+struct __attribute__((packed)) mntre_reset_firmware_version {
+	uint32_t version;
+};
```
Comment (hughsie) [inline plugins/mntre/mntre_reset.h:32]: just a single uint32 or you're expecting other stuff in the future? If the former, then maybe just remove this struct, and if the latter it should use the .rs parsing stuff -- yell if you need help.

File: plugins/mntre/fu-mntre-device.c (hunk 2 above)
Comment (zeha) [reply plugins/mntre/fu-mntre-device.c:32]: we need to discover the "2". have to see if that can be done in __init.

File: plugins/mntre/fu-mntre-device.c (hunk 4 above)
Comment (zeha) [reply plugins/mntre/fu-mntre-device.c:123]: i remember there was a problem with having it in `_init`, but will check

File: plugins/mntre/fu-mntre-device.c (hunk 2 above)
Comment (hughsie) [reply plugins/mntre/fu-mntre-device.c:32]: Aha, I've added this in probe to find it via the class/subclass/protocol.

File: plugins/mntre/fu-mntre-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/mntre/fu-mntre-device.c:117]: I fixed this.

File: plugins/mntre/fu-mntre-device.c (hunk 5 above)
Comment (hughsie) [reply plugins/mntre/fu-mntre-device.c:122]: I fixed this.

File: plugins/mntre/mntre_reset.h (hunk 7 above)
Comment (hughsie) [reply plugins/mntre/mntre_reset.h:32]: I added the rs file -- this will come in handy for the future i think

**Comment by hughsie** [comment]
@zeha I've tried this myself using a random Raspberry Pi Pico i had here with your 00000527 firmware -- but I can't get the reboot `Bootsel` working -- the reboot `Reset` works, as does the `GetFirmwareVersion` -- and I'm wondering if there's any hw mod you do on the Raspberry Pi Pico to allow jumping into the uf2 mode from the runtime.

**Comment by hughsie** [comment]
@zeha okay, I got curious; I built https://source.mnt.re/zeha/pocket-reform/-/tree/zeha-sysctl-usbids and that seems to reboot to BOOTSEL just fine -- but that doesn't have the GetFirmwareVersion call.

Note; if you wanted to make this much simpler you could just use the bcdREV of the runtime, rather than define the non-standard GetFirmwareVersion=1 control request..

**Comment by zeha** [comment]
> @zeha okay, I got curious; I built https://source.mnt.re/zeha/pocket-reform/-/tree/zeha-sysctl-usbids and that seems to reboot to BOOTSEL just fine -- but that doesn't have the GetFirmwareVersion call.

Sorry, for "reasons" I'm working on this branch: https://source.mnt.re/zeha/pocket-reform/-/tree/zeha-usbpd

> Note; if you wanted to make this much simpler you could just use the bcdREV of the runtime, rather than define the non-standard GetFirmwareVersion=1 control request..

I'm confused about bcdREV. Do you mean bcdDevice? The hardware needs to announce incompatible revisions in some way, so the plan was to use bcdDevice to indicate those.

I don't like inventing a new GetFirmwareVersion request, so I'd take better options. However I still expect we'll have other non-standard interface commands there.

One thing we probably want is to flag a running firmware as a "custom build" made by the device owner, which should not be updated. My vague idea was to expose that as another flag in GetFirmwareVersion or a second request.

File: plugins/mntre/mntre_reset.h (hunk 7 above)
Comment (zeha) [reply plugins/mntre/mntre_reset.h:32]: This .h was intended to be shared between firmware and fwupd and any other interested code.

**Comment by zeha** [comment]
> > I'm confused about bcdREV. Do you mean bcdDevice?
> 
> Yup, sorry, too many acronyms. Using the bcdDevice GUID for the firmware "stream" is quite sensible.

Thanks for confirming this plan!

> > I don't like inventing a new GetFirmwareVersion request, so I'd take better options.
> 
> What about an additional [BOS DS-20 descriptor](https://fwupd.github.io/libfwupdplugin/ds20.html)? You could just have `Version=00000527` and then it would just work. An even more galaxy-brain idea would be to include a much bigger entry in the descriptor, e.g.
> 
> ```
> Vendor=MNT Research
> Plugin=rp2_pico
> Flags=internal
> Version=00000527
> VersionFormat=number
> Icon=computer
> CounterpartGuid=BLOCK\VEN_2E8A&DEV_0003
> ```

I like this a lot. Will try it.

> ...and then rename the `mntre` plugin to `rp2-pico` (or something like that) as the `Reset` interface seems to be common to a lot of other hardware, not just MNT.

Theoretically yes, but the default implementation does not reply to the Reset command; I fixed that.
Also I'm not sure how well-specified the interface is and/or if RPI intends to keep this interface stable.

>  Crazy?

No :-)

**Comment by hughsie** [comment]
> Theoretically yes, but the default implementation does not reply to the Reset command; I fixed that.

I don't think it's too unusual to just reset TBH; other plugins just cope with the pipe error and return success.

I actually had a play with the DS-20 idea, and this is what I came up with -- I think to do properly some stuff needs to go up into tinyusb upstream, but in the meantime it's good enough to play with:

     diff --git a/pocket-reform-sysctl-fw/src/usb_descriptors.c b/pocket-reform-sysctl-fw/src/usb_descriptors.c
     index ba51da9..30b4c19 100644
     --- a/pocket-reform-sysctl-fw/src/usb_descriptors.c
     +++ b/pocket-reform-sysctl-fw/src/usb_descriptors.c
     @@ -97,7 +97,7 @@
      static const tusb_desc_device_t usbd_desc_device = {
          .bLength = sizeof(tusb_desc_device_t),
          .bDescriptorType = TUSB_DESC_DEVICE,
     -    .bcdUSB = 0x0200,
     +    .bcdUSB = 0x0210, // needed to export a BOS descriptor
          .bDeviceClass = TUSB_CLASS_MISC,
          .bDeviceSubClass = MISC_SUBCLASS_COMMON,
          .bDeviceProtocol = MISC_PROTOCOL_IAD,
     @@ -139,6 +139,59 @@ static const char *const usbd_desc_str[] = {
      #endif
      };
      
     +//------------- DS-20 -------------//
     +#define TUD_BOS_DS_20_DESC_LEN   28
     +
     +#define BOS_TOTAL_LEN      (TUD_BOS_DESC_LEN + TUD_BOS_DS_20_DESC_LEN)
     +
     +#define DS_20_DESC_LEN  32
     +
     +#define TUD_BOS_DS20_UUID   \
     +  0x63, 0xec, 0x0a, 0x01, 0x74, 0xf5, 0xcd, 0x52, \
     +  0x9d, 0xda, 0x28, 0x52, 0x55, 0x0d, 0x94, 0xf0
     +
     +#define TUD_BOS_DS20_DESCRIPTOR(_desc_set_len, _vendor_code) \
     +  TUD_BOS_PLATFORM_DESCRIPTOR(TUD_BOS_DS20_UUID, U32_TO_U8S_LE(0x0001090e), U16_TO_U8S_LE(_desc_set_len), _vendor_code, 0)
     +
     +uint8_t const desc_bos[] =
     +{
     +    // total length, number of device caps
     +    TUD_BOS_DESCRIPTOR(BOS_TOTAL_LEN, 1),
     +
     +    // DS-20, as used in fwupd
     +    TUD_BOS_DS20_DESCRIPTOR(DS_20_DESC_LEN, 0x42)
     +};
     +
     +const uint8_t *tud_descriptor_bos_cb(void)
     +{
     +    return desc_bos;
     +}
     +
     +static const uint8_t desc_ds20[] =
     +{
     +    0x50, 0x6c, 0x75, 0x67, 0x69, 0x6e, 0x3d, 0x64, 0x66, 0x75, 0x0a, \
     +    0x49, 0x63, 0x6f, 0x6e, 0x3d, 0x63, 0x6f, 0x6d, 0x70, 0x75, 0x74, \
     +    0x65, 0x72, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 // output of ./contrib/generate-ds20.py foo.quirk -b 32
     +};
     +
     +TU_VERIFY_STATIC(sizeof(desc_ds20) == DS_20_DESC_LEN, "Incorrect size");
     +
     +bool tud_vendor_control_xfer_cb(uint8_t rhport, uint8_t stage, tusb_control_request_t const * request)
     +{
     +    // nothing to with DATA & ACK stage
     +    if (stage != CONTROL_STAGE_SETUP) return true;
     +
     +    if (request->bRequest == 0x42 && request->wIndex == 7) {
     +        // Get DS-20 descriptor
     +        return tud_control_xfer(rhport, request, (void*)(uintptr_t) desc_ds20, sizeof(desc_ds20));
     +    } else {
     +        return false;
     +    }
     +
     +    // stall unknown request
     +    return false;
     +}
     +
      const uint8_t *tud_descriptor_device_cb(void) {
          return (const uint8_t *)&usbd_desc_device;
      }

**Comment by hughsie** [comment]
@zeha how did you get on? Any complain if I rename the plugin in your branch to `rp2-pico`?

**Comment by zeha** [comment]
> @zeha how did you get on? Any complain if I rename the plugin in your branch to `rp2-pico`?

@hughsie sorry, was swamped with other work.
There's still some discussion ob how to deal with different firmware versions, but i think generally for rp2040 (and probably rp2350) devices the plugin should be fine.
It will need ignoring of the disconnect error ob switching to the bootrom, for the generic implementation.

**Comment by hughsie** [comment]
> but i think generally for rp2040 (and probably rp2350) devices the plugin should be fine

Bleh, that makes choosing a plugin name much harder :) Is `rp2-pico` not accurate or should be be something like `raspberrypi-rp2040`?

**Comment by zeha** [comment]
> > but i think generally for rp2040 (and probably rp2350) devices the plugin should be fine
> 
> Bleh, that makes choosing a plugin name much harder :) Is `rp2-pico` not accurate or should be be something like `raspberrypi-rp2040`?

Good question. I think I would go with `rp-pico`. RPI explains their naming scheme like this:

```
The post-fix numeral on RP2040 comes from the following,
1. Number of processor cores (2)
2. Loosely which type of processor (M0+)
3. floor(log2(RAM / 16kB))
4. floor(log2(nonvolatile / 128kB)) or 0 if no onboard nonvolatile storage
```

The RP2350 datasheet does not have an explanation on the name. I assume they might very well come up with a Pico 4xxx when adding two more cores.

**Comment by zeha** [comment]
I've pushed this branch to be:
* code from https://github.com/fwupd/fwupd/pull/7936
* new plugin rp-pico

Three things open in particular:
* testing that this really works
* fix the TODO to discover the USB interface by name, and not by ID
* figure out how to set FU_DEVICE_PRIVATE_FLAG_ADD_INSTANCE_ID_REV from DS-20

**Comment by hughsie** [comment]
>     testing that this really works

Did you get the BOS descriptor addition working?

>    fix the TODO to discover the USB interface by name, and not by ID

I have code for that, give me 2 mins.

> figure out how to set FU_DEVICE_PRIVATE_FLAG_ADD_INSTANCE_ID_REV from DS-20

I don't _think_ you can. The USB baseclass probe has already long been run when the DS20 gets parsed.

>    fourth item: ignore error from reset request if not using a fixed usb stack on the device

Does it bubble up as `FWUPD_ERROR_READ`?

**Comment by zeha** [comment]
> > ```
> > testing that this really works
> > ```
> 
> Did you get the BOS descriptor addition working?

Currently on it, but looking good:

```
(fwupd) 130 18:31 ch@zeppelin:~/fwupd (git)-[zeha/mntre] % fwupdtool get-devices
Loading…                 [***************                        ]17:31:58.865 FuDevice             00000527 is not a valid bcd (guessed number)
Loading…                 [************************************** ]
WARNING: This package has not been validated, it may not work properly.
mntre MNT Pocket Reform with RCORE RK3588 Module
│
├─Pocket Reform System Controller 1.0:
│     Device ID:          3ead601f43028c79656941fc3832e6b67412e2b8
│     Current version:    00000527
│     Vendor:             MNT Research (USB:0x1209)
│     Serial Number:      DE63544193253234
│     GUIDs:              df0c9fc8-88d6-5100-ba66-0f5c436fc522 ← USB\VID_1209&PID_6D07
│                         2a6ade2b-8807-5467-9019-afb61e3df415 ← USB\VID_1209&PID_6D07&REV_0001
│     Device Flags:       • Internal device
│                         • Updatable
│                         • Unsigned Payload
│                         • Can tag for emulation
...
```

> > fix the TODO to discover the USB interface by name, and not by ID
> 
> I have code for that, give me 2 mins.

Nice, thanks.

> > figure out how to set FU_DEVICE_PRIVATE_FLAG_ADD_INSTANCE_ID_REV from DS-20
> 
> I don't _think_ you can. The USB baseclass probe has already long been run when the DS20 gets parsed.

Right. Maybe we won't need it.

> > fourth item: ignore error from reset request if not using a fixed usb stack on the device
> 
> Does it bubble up as `FWUPD_ERROR_READ`?

Checking, give me a moment.

**Comment by zeha** [comment]
> > > fourth item: ignore error from reset request if not using a fixed usb stack on the device
> > 
> > 
> > Does it bubble up as `FWUPD_ERROR_READ`?
> 
> Checking, give me a moment.

It does. And apparently I already wrote the code to ignore it. So this part is done!

**Comment by zeha** [comment]
@hughsie okay, this branch (incl your commit e4130bda95e1b585d6a8981058264afc5e145947) seems to work perfectly. 

How do we go from here?

I guess #7936 needs to be merged first, and then I'll squash the two remaining commits here?

**Comment by hughsie** [comment]
Did you set the version format in the descriptor before you set the version? Also, if you can rebase squashing my fixup I think we can just merge this pr for simplicity.

**Comment by zeha** [comment]
> Did you set the version format in the descriptor before you set the version? Also, if you can rebase squashing my fixup I think we can just merge this pr for simplicity.

Right, that works. I was already thinking putting VersionFormat before Version might help.

I've rebased + force-pushed this branch.

**Comment by zeha** [comment]
For reference, I tested this with firmware from https://source.mnt.re/reform/pocket-reform/-/merge_requests/21

**Comment by hughsie** [comment]
@zeha can you upload that firmware to the LVFS? it doesn't have to be public, just to the private remote. Then we can add that emulation data to that firmware and create a device test. Yell if you need a template metainfo.xml file.

**Review by hughsie — APPROVED** [review]

**Comment by zeha** [comment]
> @zeha can you upload that firmware to the LVFS? it doesn't have to be public, just to the private remote. Then we can add that emulation data to that firmware and create a device test. Yell if you need a template metainfo.xml file.

I don't have access to lvfs, but have asked @mntmn to upload the cab.

**Pull request merged**